### PR TITLE
[FLINK-16313][state-processor-api] Properly dispose of native resources when closing input split

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

FLINK-15014 refactored the `KeyedStateInputFormat` to be more composable. The change did not properly account for the new ownership of the `StateBackend` object and did not dispose of native resources. 

This fix needs to be backported to 1.10

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by sjwiesman** [comment]
cc @rmetzger

**Comment by sjwiesman** [comment]
cc @tzulitai

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java [hunk 1]
```diff
@@ -22,12 +22,9 @@
 import org.apache.flink.runtime.state.StateBackend;
 import org.apache.flink.runtime.state.memory.MemoryStateBackend;
 
-import org.junit.Ignore;
-
 /**
  * IT Case for reading state from a RocksDB keyed state backend.
  */
-@Ignore
```
Comment (tzulitai) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java]: How is this test related to the change?

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/operator/StateReaderOperator.java [hunk 2]
```diff
@@ -104,7 +104,21 @@ public void open() throws Exception {
...
+
+		try {
+			FunctionUtils.closeFunction(function);
+		} catch (Exception e) {
+			// The state backend must always be closed
+			// to release native resources.
+			exception = e;
+		}
+
+		keyedStateBackend.dispose();
```
Comment (tzulitai) [inline flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/operator/StateReaderOperator.java]: how about just:

```
try {
    FunctionUtils.closeFunction(function);
} finally {
    keyedStateBackend.dispose();
}
```

?

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java (hunk 1 above)
Comment (sjwiesman) [reply flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java]: @rmetzger disabled that test yesterday so that travis would not spuriously fail other PRs. I'm re-enabling the test as part of the fix.

Comment (tzulitai) [reply flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java]: Ah, just saw the discussions in the JIRA. Got it, seems like a revert of temporary ignore.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/operator/StateReaderOperator.java (hunk 2 above)
Comment (sjwiesman) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/operator/StateReaderOperator.java]: -1 `keyedStateBackend.dispose()` throws and exception and checkstyle did not like calling a method that throws inside of a `finally` block.

**Comment by tzulitai** [comment]
+1, LGTM.

I've locally verified that the `state-processor-api` does indeed fail with the test ignored but fix not applied, and the tests no longer fail after applying the patch.
